### ISSUE TITLE
An inconsistency with requiring in MRI vs. Opal

### ISSUE THREAD

**Issue by ylluminarious** [opal/opal#778]
So, I was recently having some problems with `require` in Opal. In MRI, requiring files works like so:
`main.rb`
```
require "./file1.rb"
require "./file2.rb"
```
`file1.rb`
```
puts "file1"
```
`file2.rb`
```
puts "file2"
```
If you run the above code, you'll get `file1` and `file2` printed out. Also note that if you remove the `./` from the `require` statements, the above code will not work. But, in Opal (as of 0.7.1), using the `./` syntax does not work and requiring only works without using `./`. In Opal 0.6.x, it seems to be that you can `require` with either `./` or without it. As far as I can tell, this problem seems to be an inconsistency between MRI and Opal.

**Comment by JoshCheek** [comment]
The dot in `./` is your current working directory. In other words, wherever you were in your file system when you ran the program. I'm not sure it makes sense to have a CWD in Opal (maybe in Node, but what does that mean in the browser?).

Anyway, requiring files works by looking through the directories in the `$LOAD_PATH` aka `$:` (also, Rubygems overrides `require` to look through its gems, if the builtin require doesn't find the file). So, you can get this example to work in both by setting the current directory into the load path. The `-I` flag does this:

![screenshot 2015-03-29 00 54 00](https://cloud.githubusercontent.com/assets/77495/6884412/206c1aee-d5ae-11e4-9135-b5d3ab84c6e5.png)

**Comment by elia** [comment]
Yeah, CWD has a shim in the browser just to make most gems happy:

https://github.com/opal/opal/blob/master/opal/corelib/dir.rb

and is linked to Node's implementation when you `require 'nodejs'` from the stdlib

https://github.com/opal/opal/blob/master/stdlib/nodejs/dir.rb

(disclaimer: nodejs support is still rough and not a priority)

---

said that I think we should add a spec for that and make it pass, the semantic would be:

- _at runtime_ to join our cwd with the path and try to require it
- _at compile time_ to assume CWD == '.' thus making `require './foo'` identical to `require 'foo'`

If anyone wants to try fixing it the relevant code should be this:

https://github.com/opal/opal/blob/6f68a488a9b0e89e14bc8d34b41fad6b19a56915/opal/corelib/runtime.js#L1280-L1296
https://github.com/opal/opal/blob/6f68a488a9b0e89e14bc8d34b41fad6b19a56915/lib/opal/nodes/call.rb#L164-L203

@ylluminarious Would that work in your case?

**Comment by ylluminarious** [comment]
@JoshCheek Yeah, I understand what you're saying about how a cwd wouldn't make much sense in the browser. Also, that `-I` flag is interesting, although I wonder how to do something like that in-browser.

@elia So, as I understand it, you want to add a spec that will test if you can do either `require './foo'` or `require 'foo'` by putting the cwd in `$LOAD_PATH`. Am I correct in my interpretation of what you're saying? Also, this will work for me; I just want Opal to be able to match up with MRI as closely as possible.

**Comment by Ajedi32** [comment]
I'm with @ylluminarious on this one. let's try to stay as close to MRI as possible unless we have a good reason to deviate.

I'm pretty sure `require './foo'` means "Load the ruby file named `foo.rb`" in the working directory.

At compile time that's easy; use the working directory that the compiler is running in (or perhaps make it configurable somehow). For runtime compilation it's a bit trickier, since there isn't really a "working directory", in the browser per-say. I guess maybe "URL of the current page" would work? Not sure... In both cases, we should match the result of `Dir.pwd` (assuming I understand MRI's behavior correctly).

**Comment by ylluminate** [comment]
Agreed. We need to keep in mind the context of what the pwd / cwd would be in each case whether server based or in-browser based compilation.  Obviously the in-browser context is essentially undefined officially, but it makes sense to think of it as @Ajedi32 stated it.  I suspect we'd want to follow ES6 module importation convention as far as that goes.  I also don't think it would be bad to allow `blah.rb` and `./blah.rb` to be aliases.  Frankly some might argue that `.\blah.rb` should also work... again, no problem there, but as long as we at least match MRI as a minimum and only expand it, I don't necessarily think it's a problem...

**Comment by JoshCheek** [comment]
In all honesty, I think it's an antipattern to use relative paths in require statements. If that's what you're wanting, you should use `require_relative`, otherwise you should fix the load path. I wish that MRI enforced this by refusing to look at relative dirs, I think it only confuses people, and those who aren't confused, it's just a minor convenience.

**Comment by ylluminate** [comment]
As much as I think most of us want to agree with you @JoshCheek, I think that ultimately the ideal is staying as close to MRI as possible and then adding to (vs taking away, so to speak) if so desired.
